### DISCUSSION TITLE
What's the canonical way to type the results of `$queryRaw`?

### DISCUSSION THREAD

**Discussion by paymog** [prisma/prisma#20477] [Q&A] [answered]
I have the following prisma query

```js
  const s = await prisma.$queryRaw<{id: string}[]>`
  SELECT s.id as id FROM "Stream" s
  WHERE (project_id, name, version) IN (
    SELECT project_id, name, MAX(version)
    FROM "Stream"
    GROUP BY project_id, name
  );`
```

I've defined the `Stream` in my Prisma schema like so:

```prisma
model Stream {
  id                 String       @default(cuid())
  name               String
  version            Int
  project_id         String


  @@id([id, version])
  @@unique([project_id, name, version])
}
```

What's the canonical way to type the `$queryRaw` result using the Prisma type tools like `Prisma.validator` and all of the other generated types? I've tried this but it doesn't seem right to me:

```js
  const streamInclude = Prisma.validator<Prisma.StreamSelect>()({
    id: true,
  })
  const s = await prisma.$queryRaw<typeof streamInclude[]>`
  SELECT s.id as id
FROM "Stream" s
WHERE (project_id, name, version) IN (
    SELECT project_id, name, MAX(version)
    FROM "Stream"
    GROUP BY project_id, name
);`
```

I feel like there's a more type safe way to type the results than `{id: string}[]` but I'm not sure if that's and true, and if it is how to do it.

**Comment by ludralph** [comment, marked as answer]
Hi @paymog 👋 

You can leverage [SafeQL](https://safeql.dev/) to write raw queries in a type-safe way.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
